Base class refactoring & 'base-x' dependency

theobat opened the pull request:
Here's the PR, it can be merged for it passes the current tests smoothly but I'd like to point out a few things :
- **base-x library does not support padding**. Haven't had the chance to check what's the multibase project policy on them, but if that's a problem we might need different implementations (hopefully for a few code bases only)
- base-x project does not have the unary base support at the moment, it might be worth it to try to make a PR on their project regarding this. I'll try to get a deeper understanding of the actual encoding algorithm to see how complicated or not that would be. 
  PS: For now the unary base throws an exception at runtime stating it is not implemented yet
- Last but not least, I haven't added new tests while I added support for many other bases, perhaps we should wait to improve the test base before merging this

Of course any other remark or fix needed regarding architecture/implementation is welcome

dignifiedquire commented on package.json:
```diff
@@ -44,9 +44,11 @@
-    "bs58": "^3.0.0"
+    "base-x": "1.0.4",
+    "bs58": "^3.0.0",
+    "webpack": "^1.13.2"
```
`webpack` does not need to be a dependency. This comes in as a devDep through `aegir` already.

dignifiedquire commented on package.json:
```diff
@@ -44,9 +44,11 @@
   "dependencies": {
-    "bs58": "^3.0.0"
+    "base-x": "1.0.4",
+    "bs58": "^3.0.0",
```
do we still need `bs58` now that we use `base-x`?

theobat replied:
No, indeed we don't need it, but several test cases still rely on that package (and I didn't want to change the test cases before a first code review) I'll remove it while I refactor the existing bs58 dependent tests (and I'll add new test cases accordingly)

theobat commented:
Thanks for the review @dignifiedquire I'll add new tests and remove the pointless dependencies. Regarding the padding thing, [It seems the multibase code reference has changed to integerate this information](https://github.com/multiformats/multibase/blob/master/multibase.csv). So the remaining tasks are  :

- [ ] Correct the base codes
- [ ] Refactor bs58 tests
- [ ] Add all supported base test cases 
- [ ] Remove pointless dependencies

Why don't we just reed the multibase repo's csv by the way ? wouldn't that be better than this constant.js ? (although we'd have to add the alphabets there which might be a bit tedious for the original purpose of this file...)
The support for padding will be for another PR.

theobat commented:
@dignifiedquire this is ready to be merged (including changes from your previsou remarks).

dignifiedquire commented:
@theobat thank you! I am travelling today so it might take a day or two until I can properly review it

diasdavid commented on src/Base.js:
```diff
@@ -0,0 +1,39 @@
+
+  getAlphabet () {
+    return this.alphabet
+  }
```
No need for this methods, as they are sync and only return properties that are defined.

diasdavid commented on src/Base.js:
```diff
@@ -0,0 +1,39 @@
+'use strict'
+
+class Base {
+  constructor (name, code, implementation, alphabet) {
```
what defines `implementation`?

theobat replied:
- **implementation** defines the module in which the actual encoding/decoding functions are implemented. see [this](https://github.com/multiformats/js-multibase/pull/7/files#diff-a7748843fbda5167d065231e3a2e7a1fR8) in constants.js for the use case. An implementation is a class taking an alphabet as an argument because it's a key factor in the encoding/decoding process (following base-x architecture).

- The idea behind this is that several bases may have implementations from different locations/modules so it's useful to have an object (and a summary) of all of them in one location (hence the constants.js). But I'm open to any refactoring for this (although I believe the multi-sourced implementations statement should be taken into account). Also even if the unary base is deployed in the base-x module (it is missing in base-x right now, which explains the empty string in the base1 list of arguments in constants.js), **given the very nature of the base-x encoding/decoding algorithm in base-x his author seems to imply that there is no possible support for padded encodings** (and indeed, after a code lookup it seems to be only bitwise manipulation on the input yielding the new base representation.

diasdavid replied:
- [ ] @theobat thank you for the throughout explanation, can you make that be part of code comments and/or readme so that it is clear to everyone?

diasdavid commented on src/multibase.js:
```diff
@@ -53,16 +48,13 @@ function decode (bufOrString) {
   }
-
-  return decode(bufOrString)
+  return decodeObject.data
```
Looks like line 51~57 can be reduced to:

`return new Buffer(base.decode(bufOrString.toString()))`

theobat replied:
I'm fine with that, although I'm not sure this is complying with the comment above : 
```
// receives a buffer or string encoded with multibase header
// decodes it and returns an object with the decoded buffer
// and the encoded type { base: <name>, data: <buffer> }ments made above the function :
```
I didn't want to change the behavior but I feel the expected output in the test is wrong and this comment is right (look at the GO implementation on this : [here](https://github.com/multiformats/go-multibase/blob/master/multibase.go#L36) (couldn't find any actual spec regarding this)

diasdavid replied:
👍

diasdavid commented on src/multibase.js:
```diff
@@ -72,56 +64,32 @@ function isEncoded (bufOrString) {
   }
 
-  return name
+  return base
```
getBase should be different from getCode as in getCode is the code used to identify the base and getBase is the string that is the name of the base

diasdavid commented on test/multibase.spec.js:
```diff
@@ -13,7 +31,7 @@ describe('multibase', () => {
     it('fails on no buf', () => {
       expect(() => {
-        multibase('base58')
+        multibase('base1')
```
why this change?

theobat replied:
`base58` doesn't exist anymore amongst the constant.js names, I thought it'd be less clear to let it that way.
(like -are we actually testing that it fails on the absence of buffer or is it failing before because of the non existing name ? turns out to be the same functionally speaking, no matter what the first argument is it will fail for the good reason)

diasdavid commented:
Thank you @theobat, this is looking good :) Submitted some CR comments for you.

theobat commented:
@diasdavid thanks for the CR, here's what I changed following your advices:
- removed `getName`, `getCode` and `getAlphabet` methods, replaced their usage by base.name and base.code
- Added instructions to the Readme to add new bases or change existing implementations

Here's whate remains unclear to me (and hence what I did **not** change - yet) :

- Some getBase change I do not understand  in `multiformats.js`:
> getBase should be different from getCode as in getCode is the code used to identify the base and getBase qis the string that is the name of the base


- `decode` function in multiformats.js :
```js
 // receives a buffer or string encoded with multibase header
 // decodes it and returns an object with the decoded buffer
 // and the encoded type { base: <name>, data: <buffer> }
// from @theobat : This is not what the multibase.spec.js test is waiting for,
// hence the return decodeObject.data
 function decode (bufOrString) {
```
Depending on the expected output of this, either I do as you said (removing everything but the `return new Buffer(base.decode(bufOrString.toString()))` **or** I change the faulty test.

theobat commented:
Can this be merged ? or can someone clarify the remaining changes ? (not urgent but I'd like to make the changes -if any- while I still have everything in mind)

diasdavid commented on README.md:
````diff
@@ -83,6 +83,58 @@ Loading this module through a script tag will make the ```Multibase``` obj avail
+    ```js
+    const supportedBases = [
+      // ... ['your-base-name']
+    ```
````
awesome!

diasdavid commented:
@theobat just this last one and good to go :)
https://github.com/multiformats/js-multibase/pull/7#discussion_r86859392

diasdavid commented:
🎉🎉🎉🎉🎉🎉🎉
Thank you @theobat :)
